File: linked-list/05-delete-node-at-pos.ts
// import classes and types
import { dummy_llist } from "./utils/mock";
import { LinkedListNode } from "./utils/types";

// Function : to delete node at any position
const delete_node_at_pos = (head: null | LinkedListNode, pos: number): null | LinkedListNode => {
    if (head === null || pos < 0) {
        return head;
    };
    if (pos === 0) {
        head = head?.next || null;
        return head;
    }
    let curr_node = head;
    let curr_position = 0;
    while (curr_position !== pos - 1 && curr_node.next) {
        curr_position++;
        curr_node = curr_node.next;
    };
    curr_node.next = curr_node.next?.next || null;
    return head;
};

// print initial value of llist
console.log("BEFORE MANIPULATIONS");
dummy_llist.print();

// delete nodes at random positions
console.log("AFTER MANIPULATIONS");
dummy_llist.head = delete_node_at_pos(dummy_llist.head, 2);
dummy_llist.print();

dummy_llist.head = delete_node_at_pos(dummy_llist.head, 3);
dummy_llist.print();

dummy_llist.head = delete_node_at_pos(dummy_llist.head, 0);
dummy_llist.print();

// testing : position = -1 (underflow)
dummy_llist.head = delete_node_at_pos(dummy_llist.head, -1);
dummy_llist.print();

// testing : position = 500 (overflow)
dummy_llist.head = delete_node_at_pos(dummy_llist.head, 500);
dummy_llist.print();

// OUTPUT
// BEFORE MANIPULATIONS
// {"data":1,"next":{"data":2,"next":{"data":3,"next":{"data":4,"next":{"data":5,"next":null}}}}}
// AFTER MANIPULATIONS
// {"data":1,"next":{"data":2,"next":{"data":4,"next":{"data":5,"next":null}}}}
// {"data":1,"next":{"data":2,"next":{"data":4,"next":null}}}
// {"data":2,"next":{"data":4,"next":null}}
// {"data":2,"next":{"data":4,"next":null}}
// {"data":2,"next":{"data":4,"next":null}}

File: linked-list/05-delete-node-at-pos.py
# ==============================================================================================

class LinkedList():  # class LinkedList
    def __init__(self):
        self.head = None

    # method to insert node
    def insert_node(self, data: int):
        if(self.head == None):
            self.head = Node(data)
            return
        curr_node = self.head
        while(curr_node.next):
            curr_node = curr_node.next
        curr_node.next = Node(data)

    # method to print node
    def print_list(self):
        print('The Linked List', end=" ")
        curr_node = self.head
        while(curr_node):
            print("->", curr_node.data, end="")
            curr_node = curr_node.next
        print()


class Node:  # class Node
    def __init__(self, data):
        self.data = data
        self.next = None


# Function: to delete node at any position
def delete_node_at_pos(head: None | Node, position: int) -> None | Node:
    if(head == None or position < 0):
        return head
    if(position == 0):
        return head.next
    curr_node = head
    curr_position = 0
    while(curr_position != position - 1 and curr_node):
        curr_position += 1
        curr_node = curr_node.next
    if(curr_node is None):
        return head
    curr_node.next = curr_node.next.next
    return head


# create a linked list
my_llist = LinkedList()

# inserting elements to fill the list
my_llist.insert_node(10)
my_llist.insert_node(20)
my_llist.insert_node(30)
my_llist.insert_node(40)
my_llist.insert_node(50)

# printing list before manipulations
print('BEFORE MANIPULATIONS')
my_llist.print_list()

print()

# update elements at any position
print('AFTER MANIPULATIONS')
my_llist.head = delete_node_at_pos(my_llist.head, 3)
my_llist.print_list()

my_llist.head = delete_node_at_pos(my_llist.head, 0)
my_llist.print_list()

my_llist.head = delete_node_at_pos(my_llist.head, 1)
my_llist.print_list()

# testing: position = -100 (underflow)
my_llist.head = delete_node_at_pos(my_llist.head, -100)
my_llist.print_list()

# testing: position = 100 (overflow)
my_llist.head = delete_node_at_pos(my_llist.head, 100)
my_llist.print_list()

# OUTPUT
# BEFORE MANIPULATIONS
# The Linked List -> 10-> 20-> 30-> 40-> 50

# AFTER MANIPULATIONS
# The Linked List -> 10-> 20-> 30-> 50
# The Linked List -> 20-> 30-> 50
# The Linked List -> 20-> 50
# The Linked List -> 20-> 50
# The Linked List -> 20-> 50
